default/AI/ResearchAI.py

```diff
@@ -20,21 +32,402 @@ def get_research_index():
+        elif item.type == fo.unlockableItemType.shipHull:
+            unlocked_hulls.append(item.name)
+    if not (unlocked_parts or unlocked_hulls):
+        return 0
+    old_designs = ship_designer.optimize_design(consider_fleet_count=False)
+    new_designs = ship_designer.optimize_design(additional_hulls=unlocked_hulls, additional_parts=unlocked_parts, consider_fleet_count=False)
+    if not (old_designs and new_designs):
+        # AI is likely defeated; don't bother with logging error message
+        return 0
+    old_rating, old_pid, old_design_id, old_cost = old_designs[0]
+    old_design = fo.getShipDesign(old_design_id)
+    old_rating = old_rating
+    new_rating, new_pid, new_design_id, new_cost = new_designs[0]
+    new_design = fo.getShipDesign(new_design_id)
+    new_rating = new_rating
+    if new_rating > old_rating:
+        ratio = (new_rating - old_rating) / (new_rating + old_rating)
+        return ratio * 1.5 + 0.5
+    else:
+        return 0
+
+def get_population_boost_priority(tech_name=""):
+    return 2
+
+def get_stealth_priority(tech_name=""):
+    max_stealth_species = get_max_stealth_species()
+    if max_stealth_species[1] > 0:
+        print "Has a stealthy species %s. Increase stealth tech priority" % max_stealth_species[0]
+        return 1.5
+    else:
+        return 0.1
+
+def get_xeno_genetics_priority():
+    if foAI.foAIstate.aggression < fo.aggression.cautious:
+        return get_population_boost_priority()
+    if has_only_bad_colonizers():
+        # Empire only have lousy colonisers, xeno-genetics are really important for them
+        print "Empire has only lousy colonizers, increase priority to xeno_genetics"
+        return get_population_boost_priority() * 3
+    else:
+        # TODO: assess number of planets with Adequate/Poor planets owned or considered for colonies
+        return 0.6 * get_population_boost_priority()
+
+def get_artificial_black_hole_priority():
+    if has_star(fo.starType.blackHole) or not has_star(fo.starType.red):
+        print "Already have black hole, or does not have a red star to turn to black hole. Skipping ART_BLACK_HOLE"
+        return 0
+    for tech in AIDependencies.SHIP_TECHS_REQUIRING_BLACK_HOLE:
+        if tech_is_complete(tech):
+            print "Solar hull is researched, needs a black hole to produce it. Research ART_BLACK_HOLE now!"
+            return 999
+    return 1
+
+def get_hull_priority(tech_name):
+    hull = 1
+    offtrack_hull = 0.05
+
+    chosen_hull = choices['hull']
+    organic = hull if chosen_hull % 2 == 0 or choices['extra_organic_hull'] else offtrack_hull
+    robotic = hull if chosen_hull % 2 == 1 or choices['extra_robotic_hull'] else offtrack_hull
+    if ColonisationAI.got_ast:
+        extra = choices['extra_asteroid_hull']
+        asteroid = hull if chosen_hull == 2 or extra else offtrack_hull
+        if asteroid == hull and not extra:
+            organic = offtrack_hull
+            robotic = offtrack_hull
```

Comment:
I think this may potentially screw up our game if we are fairly advanced in a hull line and then suddenly asteroids become available and we prefer to restart from scratch...

Not sure how to tackle this easily, but probably should be adressed in the future.